### ISSUE TITLE
Kubelet: We should define Container Waiting state better.

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#21125]
In Kubernetes, there are 3 `ContainerState`: `Running`, `Waiting` and `Terminated`. Among them, `Running` is pretty straight forward, but `Waiting` and `Terminated` are a bit ambiguous, especially `Waiting`. Even in the [official document](http://kubernetes.io/v1.1/docs/user-guide/pod-states.html), there is no specific definition for these 3 states.

For user, they need to know what these states mean, so they could inspect the system better.
For developer, the ambiguity really makes the [code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L3239) messy and buggy. 

In the current implementation, container will be considered to be `Waiting` in 3 situations:
* The container is to be started for the first time but hasn't been started yet.
* The container is in backoff crash loop.
* All historic containers are garbage collected, the container is treated as a newly started one. (This is a bug, not what we really expect)

Is this exactly what `Waiting` means? If so, we should document it and cleanup the code. If not, we should define it.

There are some possible definitions:
1) Only before container is successfully started **for the first time**, it is `Waiting`.
The lifecycle of a container should be:
```
Waiting -> Running -> Terminated -> Running -> Terminated -> ... -> Terminated.
```
2) **Each time**, after container is tried to be restarted in `SyncPod()` but before it is observed by `PLEG`, it is `Waiting`. 
The lifecycle of a container should be:
```
Waiting -> Running -> Terminated -> Waiting -> Running -> ... -> Terminated.
```
3) Only if the container will never be restarted again, it is `Terminated`, otherwise it is `Waiting`.
The lifecycle of a container should be:
```
Waiting -> Running -> Waiting -> Running -> ... -> Terminated.
```

I think option 2) should be better.

/cc @kubernetes/sig-node Any thoughts?

Also Ref #17971

**Comment by vishh** [comment]
The `Terminated` state is required to convey [termination reason](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/v1/types.go#L1167). 
Like you mentioned, option `2` makes sense and isn't that the current behavior? Kubelet's current implementation doesn't matter to the end user, as long as the semantic meaning does not change for each state. Excepting fixing `All historic containers are garbage collected, the container is treated as a newly started one` are there any other issues?

**Comment by yujuhong** [comment]
> The Terminated state is required to convey termination reason.

We can always convey this in the LastTerminationState

> Like you mentioned, option 2 makes sense and isn't that the current behavior? Kubelet's current implementation doesn't matter to the end user, as long as the semantic meaning does not change for each state. Excepting fixing All historic containers are garbage collected, the container is treated as a newly started one are there any other issues?

The problem is, we never clearly state when a container should be in the waiting state, and there are some (seemingly) inconsistency in our code (e.g., #17971). It helps to clarify the container lifecycle and properly document it.

**Comment by Random-Liu** [comment]
@vishh If option 2 is what we want, now we only set container with `ErrCrashLoopBackOff` to `Waiting`, but not for other possible errors occur during restarting.

**Comment by Random-Liu** [comment]
Status update: Offline discussed with @dchen1107 and @bgrant0607 about this today. @bgrant0607 also thinks 2) is the way we should go.

**Comment by Random-Liu** [comment]
This should have been fixed in #21224. Close this one.
